protocol-designer/src/utils/index.ts

```diff
@@ -357,52 +361,89 @@ export const getMaxConditioningVolume = (args: {
   )
 }
 
-// for stacking
+//  for stacking
 export function getLocationStackTopToBottom(
   labwareId: string,
   labwareLocationUpdate: Record<string, string>,
-  moduleLocationUpdate: Record<string, string>
+  moduleLocationUpdate: Record<string, string>,
+  moduleEntities: ModuleEntities
 ): string[] {
-  const stack = []
+  const stack: string[] = []
   let current = labwareId
 
-  //  while parent still exists
   while (true) {
     stack.push(current)
-    const parent =
-      labwareLocationUpdate[current] || moduleLocationUpdate[current]
+
+    const slot = labwareLocationUpdate[current]
+    const parent = slot ?? moduleLocationUpdate[current]
+    const isOnHopper =
+      moduleEntities[slot] != null &&
+      moduleEntities[slot].type === FLEX_STACKER_MODULE_TYPE
+
+    if (isOnHopper) {
+      // Hopper stack shape: [labware, hopper, moduleId, slot]
+      // So when the node is on the hopper, insert the hopper marker once
+      stack.push(HOPPER_STACKER_LOCATION)
+    }
+
     if (!parent) break
+
     current = parent
   }
-
   return stack
```

Comment:
Ah I hadn't noticed that we had a while loop that relies on a break statement to exit! That scares me a bit. Maybe we can implement cycle detection to make sure we don't accidentally end up in an infinite while loop
```suggestion
  const stack: string[] = []
  const visited = new Set<string>()
  let current: string | undefined = labwareId

  while (current != null) {
    // Cycle detection: if we've seen this node before, break to prevent infinite loop
    if (visited.has(current)) {
      break
    }
    visited.add(current)
    stack.push(current)

    const slot: string | undefined = labwareLocationUpdate[current]
    const parent: string | undefined = slot ?? moduleLocationUpdate[current]
    const isOnHopper =
      moduleEntities[slot] != null &&
      moduleEntities[slot].type === FLEX_STACKER_MODULE_TYPE

    if (isOnHopper) {
      // Hopper stack shape: [labware, hopper, moduleId, slot]
      // So when the node is on the hopper, insert the hopper marker once
      stack.push(HOPPER_STACKER_LOCATION)
    }

    current = parent
  }

  return stack
```

Reply:
good call out